src/storage/src/source/kafka.rs

```diff
@@ -1011,6 +1031,43 @@ fn construct_source_message(
                     .into();
                 packer.push(d)
             }
+            KafkaMetadataKind::Header { key, use_bytes } => {
+                match msg.headers() {
+                    Some(headers) => {
+                        let d = headers
+                            .iter()
+                            .filter(|header| header.key == key)
+                            .last()
+                            .map(|header| match header.value {
+                                Some(v) => {
+                                    if *use_bytes {
+                                        Ok(Datum::Bytes(v))
+                                    } else {
+                                        match str::from_utf8(v) {
+                                            Ok(str) => Ok(Datum::String(str)),
+                                            Err(_) => Err(DecodeError {
+                                                kind: DecodeErrorKind::Text(format!(
+                                                    "Found ill-formed byte sequence in header '{}' that cannot be decoded as valid utf-8",
+                                                    key
+                                                )),
+                                                raw: v.to_vec(),
+                                            }),
+                                        }
+                                    }
+                                }
+                                None => Ok(Datum::Null),
```

Comment:
Raising one last(hopefully) design question: Our current implementation loses information and can't distinguish between "I set the header to the value NULL" versus "The header was missing". Is this a case users care about?

Reply:
That's a good call out. The reason I've implemented it this way is that it matches the current suggested method of working with headers: https://materialize.com/docs/sql/create-source/kafka/#headers.
> Messages without a client_id header result in null values ("\N") for the parsed attribute.

I think of `INCLUDE HEADER` as a convenient (and more efficient) way of handling Kafka headers. People can always use `INCLUDE HEADERS` and manually work through these more challenging use cases. Although I'm not sure how easy it is to detect a missing header from the list of `(key text, value bytea)`.

If we want to have a more conservative version, I can change the semantics so that the source is bricked if there is a message in the stream that is missing a header referenced by `INCLUDE HEADER`. That shouldn't be a hard change to make and I'm not feeling strongly about it.

Reply:
If we don't have a strong preference, it may actually make sense to start with the conservative option that requires headers to be present. It's much easier to relax that proper if people complain compared to going from the way things are implemented now to a stricter interpretation that is not backwards compatible.

Reply:
> If we don't have a strong preference, it may actually make sense to start with the conservative option that requires headers to be present. It's much easier to relax that proper if people complain compared to going from the way things are implemented now to a stricter interpretation that is not backwards compatible.

Agree with this take!